src/main/java/com/Crowdsourcing/shootingstars/CrowdsourcingStars.java

```diff
@@ -0,0 +1,368 @@
+	{
+		int tier = Tiers.of(event.getGameObject().getId());
+		if (tier < 0)
+		{
+			return;
+		}
+
+		if (trackedStar != null)
+		{
+			trackedStar.setGameObject(null);
+		}
+	}
+
+	@Subscribe
+	public void onChatMessage(ChatMessage event)
+	{
+		if (event.getType() != ChatMessageType.DIALOG && event.getType() != ChatMessageType.MESBOX)
+		{
+			return;
+		}
+
+		Matcher m = STAR_PROGRESS.matcher(event.getMessage());
+		if (!m.matches())
+		{
+			return;
+		}
+
+		if (trackedStar == null)
+		{
+			log.error("Tracked star not found");
+			return;
+		}
+
+		int tier = MoreObjects.firstNonNull(Ints.tryParse(m.group("tier")), -1);
+		int progress = MoreObjects.firstNonNull(Ints.tryParse(m.group("progress")), -1);
+
+		trackedStar.setTier(tier);
+		trackedStar.setProgress(progress);
+
+		submit();
+	}
+
+	private boolean shouldSubmit(double chance)
+	{
+		double n = Math.random();
+		return n >= 1d - chance;
+	}
+
+	private void submit()
+	{
+		if (trackedStar == null)
+		{
+			log.error("Missing star to submit");
+			return;
+		}
+
+		if (lastSent != null &&
+			lastSent.getTier() == trackedStar.getTier() &&
+			(lastSent.getProgress() != null && (lastSent.getProgress() == trackedStar.getProgress() || trackedStar.getProgress() < lastSent.getProgress())) &&
+			lastSent.getWorld() == trackedStar.getWorld() &&
+			lastSent.getLocation().equals(trackedStar.getLocation()))
+		{
+			log.debug("Not submitting, same information");
+			return;
+		}
+
+		lastSent = StarData.builder()
+			.tier(trackedStar.getTier())
+			.world(client.getWorld())
+			.location(trackedStar.getLocation())
+			.progress(trackedStar.getProgress() > -1 ? trackedStar.getProgress() : null)
+			.mode(HiscoreEndpoint.fromWorldTypes(client.getWorldType()))
+			.build();
+		submitStar(lastSent);
+	}
+
+	private void submitStar(StarData data)
+	{
+		String json = gson.toJson(data);
+		log.debug("submitting {}", json);
+
+		Request r = new Request.Builder()
+			.url(CROWDSOURCING_URL)
+			.post(RequestBody.create(JSON, json))
+			.build();
+
+		okHttpClient.newCall(r).enqueue(new Callback()
+		{
+			@Override
+			public void onFailure(Call call, IOException e)
+			{
+				log.debug("Error sending crowdsourcing data", e);
+			}
+
+			@Override
+			public void onResponse(Call call, Response response)
+			{
+				log.debug("Successfully sent crowdsourcing data");
+				response.close();
+			}
+		});
+	}
+
+	private void checkStarProgress()
+	{
+		if (trackedStar == null)
+		{
+			return;
+		}
+
+		if (trackedStar.getNpc() == null)
+		{
+			if (trackedStar.getGameObject() == null)
+			{
+				// star probably died
+				if (trackedStar.getTier() == 1 && trackedStar.getLocation().distanceTo2D(client.getLocalPlayer().getWorldLocation()) < 16)
+				{
+					trackedStar.setProgress(-1);
+					trackedStar.setTier(0);
+
+					submit();
+				}
+
+				reset();
+			}
+
+			return;
+		}
+
+		NPC npc = trackedStar.getNpc();
+		int progress = 100 - 100 * npc.getHealthRatio() / npc.getHealthScale();
```

Comment:
Should we just send the raw `getHealthRatio()` to the server? If `getHealthScale()` is constant for stars I think that might be less of a pain than dealing with division in java. Not sure what the value of `getHealthScale()` is for stars (seem to remember hearing 40?) but if it doesn't divide 100 evenly we also deal with int division in specific cases here.